### PR TITLE
Fix File Workaround in Disk Repo

### LINKED ISSUE #71: DiskRepo should use optimistic concurrency for CommitRef
Per Azure repos

### LINKED ISSUE #95: Fix File Workaround in DiskRepo

### PR DESCRIPTION
**Pull request by jcdickinson**
Waiting for an unlocked file will always be required. It is now async and throws correctly. The concurrency model for CommitRef is now implemented.

- Cleaning up uneeded checks in the DiskChasmRepo ctor.
- ReadObjectBatchAsync is now parallel.

Fixes #71
Fixes #95.

### CODE REVIEW

File: src/SourceCode.Chasm.Repository.Disk/DiskChasmRepo.cs [hunk 1]
```diff
@@ -23,15 +23,24 @@ public sealed partial class DiskChasmRepo : ChasmRepository
...
+
+        #endregion
+
         #region Constructors
 
         public DiskChasmRepo(string rootFolder, IChasmSerializer serializer, CompressionLevel compressionLevel, int maxDop)
             : base(serializer, compressionLevel, maxDop)
         {
             if (string.IsNullOrWhiteSpace(rootFolder) || rootFolder.Length <= 2) throw new ArgumentNullException(nameof(rootFolder)); // "C:\" is shortest permitted path
-            if (rootFolder[rootFolder.Length - 1] != Path.DirectorySeparatorChar) throw new ArgumentException("Path must end with " + Path.DirectorySeparatorChar, nameof(rootFolder));
```
Comment (grant-d) [inline src/SourceCode.Chasm.Repository.Disk/DiskChasmRepo.cs]: Why did you remove the guards

File: src/SourceCode.Chasm.Repository.Disk/DiskChasmRepo.cs [hunk 2]
```diff
@@ -42,7 +51,7 @@ public DiskChasmRepo(string rootFolder, IChasmSerializer serializer, Compression
             // Refs
             {
                 const string container = "refs";
-                var path = Path.Combine(rootPath, container) + Path.DirectorySeparatorChar;
```
Comment (grant-d) [inline src/SourceCode.Chasm.Repository.Disk/DiskChasmRepo.cs]: Appending a '/' here means we don't need to check for it every time we create a file

File: src/SourceCode.Chasm.Repository.Disk/DiskChasmRepo.cs (hunk 1 above)
Comment (jcdickinson) [reply src/SourceCode.Chasm.Repository.Disk/DiskChasmRepo.cs]: They didn't prevent any exceptions from occurring later. Everything uses `Path.Combine`.

File: src/SourceCode.Chasm.Repository.Disk/DiskChasmRepo.cs (hunk 2 above)
Comment (jcdickinson) [reply src/SourceCode.Chasm.Repository.Disk/DiskChasmRepo.cs]: It's not platform agnostic. `Path.Combine` is the only way paths should be combined.
